docs/kuneiform/procedures.md

```diff
@@ -0,0 +1,275 @@
+    }
+}
+```
+
+You can also iterate of returning clauses:
+
+```ts
+procedure iterate_over_delete() public {
+    $age int := 100;
+    for $row in delete from users
+        where age = $age
+        returning * {
+        // can access $row.id, $row.name, and $row.age
+        // for all deleted users
+    }
+}
+```
+
+Procedures that return tables can also be iterated over:
+
+```ts
+procedure get_users() public view returns table(id uuid, name text, age int) {
+    // ...
+}
+
+procedure iterate_over_users() public {
+    for $row in SELECT * FROM get_users() {
+        // can access $row.id, $row.name, and $row.age
+    }
+}
+```
+
+#### Breaking From a Loop
+
+To conditionally exit a loop, you can use `break`:
+
+```ts
+procedure get_users() public {
+    $count int := 0;
+    for $row in SELECT * FROM users {
+        // do something
+
+        // break if too many records have been read
+        $count := $count + 1;
+        if $count >= 50 {
+            break
+        }
+    }
+}
+```
+
+
+
+### Returning Values
+
+#### Return Single Records
+
+To return a single record, simply use the `return` key word:
+
+```ts
+procedure get_numbers() public view returns (num1 int, num2 int) {
+    $number1 int := 1;
+    return $number1, 2;
+}
+```
+
+To return a single row from a SQL query, you should iterate over the result and return on the first iteration:
+
+```ts
+procedure get_user($id uuid) public view returns (name text, age int) {
+    for $row in SELECT name, age from users where id = $id {
+        // this will always return on the first iteration.
+        return $row.name, $row.age;
+    }
+
+    // if no iterations, then there is no matching user.
+    error('user not found');
+}
+```
+
+#### Return Table
+
+To return a table from a procedure, you can simply return any SQL query that returns data:
+
+```ts
+procedure get_users() public view returns table(id uuid, name text, age int) {
+    return select * from users;
+}
+```
+
+#### Return Next
+
+If a procedure returns a table and should to selectively return values, you can use `return next` to return a record inside of a for loop. Unlike `return`, `return next` will not exit the procedure, but instead continue. In the below example, we query a table `prices` with columns `value` and `day`, and return the net change from each day. We assume that our first day is day 1. If a day is missing, it returns an error.
+
+```ts
+procedure get_change() public returns table(day int, change int) {
+    $last_value int := 0;
+    $last_day int := 0;
+    for $row in SELECT day, value from prices order by day asc {
+        // if we received a day out of order, return error
+        if $row.day != $last_day + 1 {
+            error(format('missing day %d', $last_day+1));
+        }
+
+        // add the change to the result set
+        return next $row.day, $row.value - $last_value;
+        // update our last seen value
+        $last_value := $row.value;
+    }
+}
```

Comment:
I guess any procedure with a `return next` somewhere has some implicit "`return collected`" at the end.  I'm just noting that there's no `return` at the end, but the `for` could never call a `return next`, and the procedure has a `returns table...`.   
This is fine, just making sure.

Reply:
Yeah, there is no `return collected`. A procedure that returns a table doesn't technically have to return anything (it would just be an empty table)